iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp

```diff
@@ -0,0 +1,177 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Conversion/StandardToLLVM/ConvertStandardToLLVM.h"
+#include "mlir/Conversion/VectorToLLVM/ConvertVectorToLLVM.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+/// Converts 4x4x4 vector contraction with matmul(A_, B) semantics to AArch64
+/// inline assembly using aarch64 4-sdot instructions. Each sdot instruction
+/// performas a single matrix-vector product and to compute matmul(A, B) with
+/// matrix-vector products B is transposed.
+struct ConvertVectorContract4x4x4_i8i8i32_ToAArch64InlineAsmPattern
+    : public OpRewritePattern<vector::ContractionOp> {
+ public:
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractionOp,
+                                PatternRewriter &rewriter) const override {
+    auto lhsType = contractionOp.lhs().getType().cast<VectorType>();
+    auto rhsType = contractionOp.rhs().getType().cast<VectorType>();
+    auto accType = contractionOp.acc().getType().cast<VectorType>();
+    auto lhsShape = lhsType.getShape();
+    auto rhsShape = rhsType.getShape();
+    if (lhsShape[0] != 4 || lhsShape[1] != 4 || rhsShape[0] != 4 ||
+        rhsShape[1] != 4)
+      return failure();
+
+    Value inLhs = contractionOp.lhs();
+    Value inRhs = contractionOp.rhs();
+
+    if (accType.getElementType() != rewriter.getIntegerType(32)) {
+      return failure();
+    }
+    // TODO(ataei) : Currently linalg op lowering sign extends all operands to
+    // i32 we need to remove this when linalg -> vector.contract isn't
+    // explicitly adding sexti ops.
+    if (lhsType.getElementType() != rewriter.getIntegerType(8)) {
```

Comment:
Nit: Make a single lambda that returns `Optional<Value>` and use that for getting the value here.

```
auto get8BitValue = [&](Value v) -> Optional<Value> {
  if (auto parentOp = v.getDefiningOp<SignExtendIOp>()) {
    return parentOp.value();
  }
  return llvm::None;
}
auto lhs8Bit = get8BitValue(inLhs);
auto rhs8Bit = get8BitValue(inRhs);
if (!lhs8Bit || !rhs8Bit) return failure();
```

In any case, you should not use `cast<SignExtendIOp>(...)`, use `.getDefiningOp<SignExtendIOp>()`.

Reply:
Thanks! didn't now getDefiningOp does have casting semantics :)